### PR TITLE
UI fix: user with authorization to create in at least a category can access editor view in menuitems for other categories

### PR DESCRIPTION
**Pull request by LivioCavallo**
### Summary of Changes
Perform an additional authorization check at the category level in com_content edit view file: view.html.php .
This blocks an unauthorized user from accessing the editor in some circumstancies when he does not have "Create" permissions for that category.

### Testing Instructions
Create a new User Group: "Manage Category ONE" with 'Public' parent group.
In System / Global Configuration set Permission to allow "Manage Category ONE" group to Login into site.
Create new user 'User1' as member of this group.
Create a new article category "ONE" and set permissions so that "Manage Category ONE" group is allowed to "Create".
Create a new article category "TOP SECRET" and check that "Manage Category ONE" group has no permissions to "Create" in it.
Now create a menuitem, "Create Top Secret", to submit a new article; set Options so that default category "TOP SECRET" is default.

Login as "User1" into frontend and "Create Top Secret".

### Expected result
User is not authorized, not having rights to "Create" in "Top Secret" category.

### Actual result
User is allowed to enter the editor! He can even browse images folder.

### Notes
In J!3.4.4 (and I suppose in some later versions) the user was allowed to save the article!
In J!3.7.5 (I suppose starting from some previous versions) when he clicks on "Save" the operation is blocked with error; this mitigates the problem.
Anyway I think that accessing the editor (though the most dangerous functionalities are disabled) is not a "good thing": the user should be blocked as soon as he tries to access a resource he is not authorized to use, not later, when he tries to close (save) that resource.

### CODE REVIEW

File: components/com_content/views/form/view.html.php [hunk 1]
```diff
@@ -52,7 +52,15 @@ public function display($tpl = null)
 
 		if (empty($this->item->id))
 		{
-			$authorised = $user->authorise('core.create', 'com_content') || count($user->getAuthorisedCategories('com_content', 'core.create'));
+			if ($this->state->params->get('enable_category') == 1)
+			{
+				$catid = $app->getMenu()->getActive()->params->get('catid');
```
Comment (ggppdk) [inline components/com_content/views/form/view.html.php]: Is there a reason you are using $app->getMenu()->getActive()->params ?
For consistency you should use the $this->state->params to get 'catid'
e.g. in the line above you are using it to get 'enable_category' which is also a menu item parameter

**Comment by ggppdk** [comment]
Also i noticed that parameter name is 'Default category'
but parameter description says: 

> If set to 'Yes', this page will only let you create articles in the category selected below.

so "Default category" does not sound like best choice
instead of "Default category" a proper parameter name would be 
- Limit category
- Category Limitation
- Specific category

----

submit_new_article

**Comment by LivioCavallo** [comment]
Constrained / Bound category?

File: components/com_content/views/form/view.html.php (hunk 1 above)
Comment (LivioCavallo) [reply components/com_content/views/form/view.html.php]: No.
Right.

File: administrator/language/en-GB/en-GB.com_content.ini [hunk 2]
```diff
@@ -30,7 +30,7 @@ COM_CONTENT_CONFIG_LIST_SETTINGS_DESC="These settings apply for List Layouts Opt
 COM_CONTENT_CONFIGURATION="Articles: Options"
 COM_CONTENT_CREATE_ARTICLE_CANCEL_REDIRECT_MENU_DESC="Select the page the user will be redirected to after Canceling article submission. The default is to redirect to the same article submission page (cleaning form)."
 COM_CONTENT_CREATE_ARTICLE_CANCEL_REDIRECT_MENU_LABEL="Cancel Redirect"
-COM_CONTENT_CREATE_ARTICLE_CATEGORY_LABEL="Default Category"
+COM_CONTENT_CREATE_ARTICLE_CATEGORY_LABEL="Constrained Category"
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.com_content.ini]: Sort alpha order. Swap lines 33 and 34.

Comment (LivioCavallo) [reply administrator/language/en-GB/en-GB.com_content.ini]: Oooops. Tks

**Comment by infograf768** [comment]
Constrained is hard to translate.
"Specific category" reads better imho.

**Comment by ggppdk** [comment]
This is more an UX fix , that an security fix (e.g. the form display of custom fields of an non-accessible category are revealed),

i say UX because, imagine someone spending 10, 20 minutes to fill in the form and then clicking save, to discover that has no access to save the form, some real frustration

**Comment by LivioCavallo** [comment]
So should I change name to: "UX fix: ..."  ?
(Or programmer safety? :wink: )

**Comment by infograf768** [comment]
Hmm, found a bug, unrelated to this PR.
One can save a Create Article menu item, settting "Default Category" to Yes and NOT filling the "Choose a Category" field.
In frontend, using such a setting forces to save the article created in the category with the smallest Id.  catid param is therefore empty.

Therefore I suggest to use
``` 
$catid = $this->state->params->get('catid');
if (!empty($catid)
{
   $authorised = $user->authorise('core.create', 'com_content.category.' . $catid);
etc.
```
as I could not find a way to solve that bug.

**Comment by LivioCavallo** [comment]
Oh! Thanks. Fixed.

There is a way to add client-side field validation to params form field in backend?
In this case it would be useful something like 'showon:', let's say 'requireon:'.

**Comment by LivioCavallo** [comment]
It seems there were problems with drone and appveyor, but I don't find a way to restart them. Is it normal?
Or there is really any error?

**Comment by alikon** [comment]
i think we still need to manage properly the issue referenced by @infograf768, your fix just show `Error save not permitted`  but maybe is better to open a different issue for that

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on 175a3283a6c8025fbd982800f1ff501ec82ea93c<br /><br />as per pr testing info<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17674">issues.joomla.org/tracker/joomla-cms/17674</a>.</sub>

**Comment by LivioCavallo** [comment]
You are right @alikon.
Now we have a new issue: #17708.

**Comment by ggppdk** [comment]
@LivioCavallo 

I have made PRs against you patching both places, that 'enable_category' is used

components/com_content/views/form/view.html.php
and 
components/com_content/models/form.php

now code will work properly even if user has set 'enable_category' to YES and did not select category
- it will now behave as if 'enable_category' is NO

fixing is  #17708 is now not needed

**Comment by LivioCavallo** [comment]
Thanks @ggppdk, I like your idea, I think it's a good workaround. I merged it.
Anyway I think that the "Selected category" should be 'required' if Default category is ON and the user should be blocked on saving or at least receive an alert saying that without selecting a category there won't be any constraints on the category, as if 'enable_category' is OFF, as implemented by your PRs.
I think that issue #17708 is still needed.

File: components/com_content/views/form/view.html.php [hunk 3]
```diff
@@ -50,9 +50,19 @@ public function display($tpl = null)
 		$this->form        = $this->get('Form');
 		$this->return_page = $this->get('ReturnPage');
 
+		$authorised = false;
+
 		if (empty($this->item->id))
 		{
-			$authorised = $user->authorise('core.create', 'com_content') || count($user->getAuthorisedCategories('com_content', 'core.create'));
+			$catid = $this->state->params->get('catid');
+			if ($this->state->params->get('enable_category') == 1 && $catid)
```
Comment (infograf768) [inline components/com_content/views/form/view.html.php:57]: Please add a blank line before a conditional

**Comment by infograf768** [comment]
> I think that issue #17708 is still needed.

Agree, but can be done separately. (`requiredon` is not a simple task to implement)

**Comment by LivioCavallo** [comment]
@infograf768 : I agree. This PR fixes problems. Thanks all.
A better Long term user experience will be eventually provided by #17708

File: components/com_content/views/form/view.html.php [hunk 4]
```diff
@@ -50,9 +50,20 @@ public function display($tpl = null)
 		$this->form        = $this->get('Form');
 		$this->return_page = $this->get('ReturnPage');
 
+		$authorised = false;
```
Comment (Quy) [inline components/com_content/views/form/view.html.php]: This is not necessary and can be removed since it will be assigned in the following if statement.

**Comment by LivioCavallo** [comment]
Why is AppVeyor failing?

**Comment by LivioCavallo** [comment]
I tested it in Jommla! 3.7.4, 3.7.5 and 3.8 beta. This patch works as expected.
Did someone else test it?

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on ad148a7d590afd37c720b3fcbdaaa7b7d4a4e1f8<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17674">issues.joomla.org/tracker/joomla-cms/17674</a>.</sub>

**Comment by ghost** [comment]
@alikon can you please retest?

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on ad148a7d590afd37c720b3fcbdaaa7b7d4a4e1f8<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17674">issues.joomla.org/tracker/joomla-cms/17674</a>.</sub>

**Comment by ghost** [comment]
Status "Ready To Commit".

**Comment by richard67** [comment]
@HLeithner What happens with this PR?

**Comment by HLeithner** [comment]
Something take longer then other, thanks for this PR

**Pull request merged**